fs/fd.c

```diff
@@ -203,25 +203,36 @@ fd_t f_install(struct fd *fd, int flags) {
     return f;
 }
 
-static int fdtable_close(struct fdtable *table, fd_t f) {
+static int fdtable_close_flags(struct fdtable *table, fd_t f, int flags) {
     struct fd *fd = fdtable_get(table, f);
     if (fd == NULL)
         return _EBADF;
     if (fd->inode != NULL) // temporary hack for files like sockets that right now don't have inodes but will eventually
         file_lock_remove_owned_by(fd, table);
     int err = fd_close(fd);
     table->files[f] = NULL;
-    bit_clear(f, table->cloexec);
+    if (flags & O_CLOEXEC_)
+    	bit_set(f, table->cloexec);
+    else
+    	bit_clear(f, table->cloexec);
```

Comment:
Setting a flag on a file descriptor that you just closed doesn't make sense. This should be set by the sys_dup3 function. That would also avoid having to add all this code to the close functions.

Reply:
Thanks comment. I agree, but will get duplicate  code.
Can you add dup3? and let me learn more.

Reply:
Can you explain why it would duplicate code?